perf/src/stats.rs

```diff
@@ -0,0 +1,368 @@
+        if stream_stats.finished.load(Ordering::Relaxed) {
+            let duration = stream_stats.duration.load(Ordering::Relaxed) as u64;
+            let bps = throughput_bytes_per_second(duration, stream_stats.request_size);
+
+            if stream_stats.sender {
+                self.upload_throughput.record(bps as u64).unwrap();
+                self.upload_duration.record(duration).unwrap();
+            } else {
+                self.download_throughput.record(bps as u64).unwrap();
+                self.download_duration.record(duration).unwrap();
+                self.fbl
+                    .record(stream_stats.first_byte_latency.load(Ordering::Relaxed) as u64)
+                    .unwrap();
+                self.requests += 1;
+            }
+        }
+    }
+
+    pub fn print(&self) {
+        let dt = self.start_instant.elapsed();
+        let rps = self.requests as f64 / dt.as_secs_f64();
+
+        println!("Overall stats:");
+        println!(
+            "RPS: {:.2} ({} requests in {:4.2?})",
+            rps, self.requests, dt,
+        );
+        println!();
+
+        println!("Stream metrics:\n");
+
+        println!("      │ Upload Duration │ Download Duration | FBL        | Upload Throughput | Download Throughput");
+        println!("──────┼─────────────────┼───────────────────┼────────────┼───────────────────┼────────────────────");
+
+        let print_metric = |label: &'static str, get_metric: fn(&Histogram<u64>) -> u64| {
+            println!(
+                " {} │ {:>15} │ {:>17} │  {:>9} │ {:11.2} MiB/s │ {:13.2} MiB/s",
+                label,
+                format!(
+                    "{:.2?}",
+                    Duration::from_micros(get_metric(&self.upload_duration))
+                ),
+                format!(
+                    "{:.2?}",
+                    Duration::from_micros(get_metric(&self.download_duration))
+                ),
+                format!("{:.2?}", Duration::from_micros(get_metric(&self.fbl))),
+                get_metric(&self.upload_throughput) as f64 / 1024.0 / 1024.0,
+                get_metric(&self.download_throughput) as f64 / 1024.0 / 1024.0,
+            );
+        };
+
+        print_metric("AVG ", |hist| hist.mean() as u64);
+        print_metric("P0  ", |hist| hist.value_at_quantile(0.00));
+        print_metric("P10 ", |hist| hist.value_at_quantile(0.10));
+        print_metric("P50 ", |hist| hist.value_at_quantile(0.50));
+        print_metric("P90 ", |hist| hist.value_at_quantile(0.90));
+        print_metric("P100", |hist| hist.value_at_quantile(1.00));
+        println!();
+    }
+
+    pub fn print_json(&self) {
+        json::print(&self);
+    }
+}
+
+pub struct StreamStats {
+    id: StreamId,
+    request_size: u64,
+    bytes: AtomicUsize,
+    sender: bool,
+    finished: AtomicBool,
+    duration: AtomicU64,
+    first_byte_latency: AtomicU64,
+}
+
+impl StreamStats {
+    pub fn new(id: StreamId, request_size: u64, sender: bool) -> Self {
```

Comment:
nit: maybe instead of `sender: bool` have two different methods: `new_sender` and `new_receiver`?